packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -405,52 +434,32 @@ export class Pipeline extends PipelineBase {
-      const pipelineAndActionInDifferentRegions = pipelineAndActionRegionComparison === TokenComparison.ONE_UNRESOLVED ||
-        pipelineAndActionRegionComparison === TokenComparison.DIFFERENT;
-      if (pipelineAndActionInDifferentRegions) {
-        actionRegion = actionResource.env.region;
-
-        // if the resource is from a different stack in another region but the same account,
-        // use that stack as home for the cross-region support resources
-        const actionResourceStack = Stack.of(actionResource);
-        const actionResourceAndItsStackRegionComparison = Token.compareStrings(actionResource.env.region, actionResourceStack.region);
-        const actionResourceInSameRegionAsItsStack = actionResourceAndItsStackRegionComparison === TokenComparison.SAME ||
-          actionResourceAndItsStackRegionComparison === TokenComparison.BOTH_UNRESOLVED;
-        const pipelineAndActionResourceStackAccountComparison = Token.compareStrings(this.env.account, actionResourceStack.account);
-        const pipelineAndActionResourceStackInSameAccount = pipelineAndActionResourceStackAccountComparison === TokenComparison.SAME ||
-          pipelineAndActionResourceStackAccountComparison === TokenComparison.BOTH_UNRESOLVED;
-        if (pipelineAndActionResourceStackInSameAccount && actionResourceInSameRegionAsItsStack) {
-          otherStack = actionResourceStack;
-        }
-      }
-    } else {
-      actionRegion = action.actionProperties.region;
+    // Do a check on whether the cross-account acccess is possible given the
+    // current pipeline configuration. This has bearing on whether
+    // `artifactBucket.grantRead(action.role)` (which is going to occur down the
+    // line) can even work.
+    if (hasAccount && !sameEnvDimension(this.env.account, actionEnv.account) && !this.crossAccountKeys) {
+      throw new Error(`Pipeline must be created with 'crossAccountKeys: true' to add cross-account action '${action.actionProperties.actionName}'. Pipeline account: '${renderEnvDimension(this.env.account)}', action account: '${renderEnvDimension(actionEnv.account)}'`);
     }
 
-    // if actionRegion is undefined,
-    // it means the action is in the same region as the pipeline -
-    // so, just return the artifactBucket
-    if (!actionRegion) {
-      return {
-        artifactBucket: this.artifactBucket,
-      };
-    }
-    // get the region the Pipeline itself is in
-    const pipelineRegion = this.requireRegion();
-    // if the action is in the same region as the pipeline, nothing to do
-    if (actionRegion === pipelineRegion) {
+    // if actionRegion is undefined or the same as the pipeline region,
+    // it means the action is in the same region as the pipeline - so, just
+    // return the artifactBucket
+    if (!hasRegion || sameEnvDimension(this.env.region, actionEnv.region) || Token.isUnresolved(actionEnv.region)) {
```

Comment:
Gotta say, this conditions mystifies me a little bit. 

`hasRegion = !Token.isUnresolved(actionEnv.region)`. So `!hasRegion` == `Token.isUnresolved(actionEnv.region)`. So this condition is basically the same as `sameEnvDimension(this.env.region, actionEnv.region)`...?

Reply:
You're right that part of this is unnecessary but now you're simplifying too much.

The condition is effectively:

```
if (R || P || R) { }
```

But simplified that's still:

```
if (R || P) { }

// NOT
if (P) { }
```